### PR TITLE
Add submision conversion feature for individual submission and by module

### PR DESCRIPTION
**Pull request by muhammadfaiz12**
Late submission conversion

EDIT-767

# Description

**What?**

This feature convert a late submission to a normal submission without late penalty.

**Why?**

Teaching staff could convert a late submission

**How?**

Removing late penalties and changing the submission status

# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)

**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [ ] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [x] Manual testing.

I test the fature manually and checking wether a late submission was actually converted and get the grade without penalities.
**Did you test the changes in**

- [ ] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [ ] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [x] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?
- [x] Did you use Python type hinting in all functions that you added or edited? ([type hints](https://docs.python.org/3/library/typing.html) for function parameters and return values)

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [x] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

### CODE REVIEW

File: exercise/staff_views.py [hunk 1]
```diff
@@ -530,3 +532,61 @@ def form_valid(self, form):
     def form_invalid(self, form):
         messages.error(self.request, _('FAILURE_SAVING_CHANGES'))
         return super().form_invalid(form)
+
+class SubmissionConversion(SubmissionMixin, BaseRedirectView):
+    """
+    A POST-only view that updates a student's late or unofficial submission
+    to normal submission. Changed the status and remove the penalty
+    """
+    def post(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
```
Comment (PasiSa) [inline exercise/staff_views.py:541]: Is this missing access_mode = ACCESS.ASSISTANT?

File: exercise/staff_views.py [hunk 2]
```diff
@@ -530,3 +532,61 @@ def form_valid(self, form):
     def form_invalid(self, form):
         messages.error(self.request, _('FAILURE_SAVING_CHANGES'))
         return super().form_invalid(form)
+
+class SubmissionConversion(SubmissionMixin, BaseRedirectView):
```
Comment (PasiSa) [inline exercise/staff_views.py:536]: To follow the naming convention, I think the class name should be SubmissionConversionView, right?

File: exercise/staff_views.py [hunk 3]
```diff
@@ -530,3 +532,61 @@ def form_valid(self, form):
...
+    to normal submission. Changed the status and remove the penalty
+    """
+    def post(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
+        self.submission = self.get_submission_object()
+        self.submission.convert_late_submission()
+        self.submission.save()        
+        return self.redirect(self.submission.get_inspect_url())
+
+
+class SubmissionConversionByModule(CourseInstanceMixin, BaseRedirectView):
```
Comment (PasiSa) [inline exercise/staff_views.py:548]: Also here, the name could be SubmissionConversionByModuleView

File: exercise/templates/exercise/staff/_assessment_panel.html [hunk 4]
```diff
@@ -19,7 +19,23 @@
...
+						{% translate 'CONVERT_SUBMISSION' %}
+					</button><button
+						data-toggle="modal"
+						data-target="#resubmit-modal"
+						class="aplus-button--secondary aplus-button--sm aplus-button--right"
+						type="button"
+						title="{% translate 'HELP' %}"
+					>?</button>
+				{% endif %}
+				</form>				<button
```
Comment (PasiSa) [inline exercise/templates/exercise/staff/_assessment_panel.html:38]: line break missing here, between `</form>` and `<button>`

File: locale/en/LC_MESSAGES/django.po [hunk 5]
```diff
@@ -3711,6 +3711,10 @@ msgstr "%(points)s points required to pass the module. "
 msgid "CHANGES_ARE_POSSIBLE_IN_EXERCISES_BEFORE_MODULE_OPENING"
 msgstr "There may be changes in the assignments before the module opens!"
 
+#: exercise/templates/exercise/_user_results.html
+msgid "CONVERT_SUBMISSION_BY_MODULE"
+msgstr "Convert Late Submissions"
```
Comment (PasiSa) [inline locale/en/LC_MESSAGES/django.po:3716]: I'm not sure how much space is available here, but "Convert Late Submissions" may not be clear to user. Convert to what? Could "Approve Late Submissions" be better? Or if space is not an issue, "Convert Late Submissions to fully graded"

File: locale/en/LC_MESSAGES/django.po [hunk 6]
```diff
@@ -3838,6 +3842,10 @@ msgstr "Submission details"
 msgid "REGRADE_AUTOMATICALLY"
 msgstr "Regrade automatically"
 
+#: exercise/templates/exercise/staff/_assessment_panel.html
+msgid "CONVERT_SUBMISSION"
```
Comment (PasiSa) [inline locale/en/LC_MESSAGES/django.po:3846]: Similar comment here: meaning of "Convert Submission" may not be clear

File: locale/fi/LC_MESSAGES/django.po [hunk 7]
```diff
@@ -3721,6 +3721,10 @@ msgstr "Moduulin suorittamiseen vaaditaan %(points)s pistettä."
 msgid "CHANGES_ARE_POSSIBLE_IN_EXERCISES_BEFORE_MODULE_OPENING"
 msgstr "Tehtäviin saattaa tulla muutoksia ennen kierroksen avautumista!"
 
+#: exercise/templates/exercise/_user_results.html
+msgid "CONVERT_SUBMISSION_BY_MODULE"
+msgstr "Convert Late Submissions in Finnish"
```
Comment (PasiSa) [inline locale/fi/LC_MESSAGES/django.po:3726]: I can help with Finnish translations, once the terminology is fixed (relating to above comment on English texts)

**Review by PasiSa — COMMENTED** [review]
Here are some comments from me.

Also the commit log message could be a bit more verbose in my mind. "Submission conversion feature" is not so clear if you do not know what this relates to. After a short title line, perhaps sentence or two to clarify what the feature does would be useful. You could also add reference to the issue: "Closes #892", to link the commit with the original issue (and automatically close it).

Could you write short step-by-step instructions, for example as a follow-up comment in this pull request feed, on how to test this feature? (See recent discussion on EDIT Teams channel about improving our documentation on testing)

File: locale/en/LC_MESSAGES/django.po (hunk 5 above)
Comment (muhammadfaiz12) [reply locale/en/LC_MESSAGES/django.po:3716]: Yes, Balancing the length and descriptiveness of the name is quite hard for me. The "Convert Late Submissions to fully graded" is the most descriptive but also too long imo. Maybe i could ask for teachers opinion on the meeting on this.

File: exercise/templates/exercise/staff/_assessment_panel.html (hunk 4 above)
Comment (muhammadfaiz12) [reply exercise/templates/exercise/staff/_assessment_panel.html:38]: Thanks, will fix

File: exercise/staff_views.py (hunk 2 above)
Comment (muhammadfaiz12) [reply exercise/staff_views.py:536]: Ahh i see, well yes that makes sense. I will update that. thanks

File: exercise/staff_views.py (hunk 1 above)
Comment (muhammadfaiz12) [reply exercise/staff_views.py:541]: Well, i'm not sure, but if you think i should add that then i could do that

File: locale/fi/LC_MESSAGES/django.po (hunk 7 above)
Comment (muhammadfaiz12) [reply locale/fi/LC_MESSAGES/django.po:3726]: Yes, thanks for the help

File: exercise/staff_views.py [hunk 8]
```diff
@@ -530,3 +532,73 @@ def form_valid(self, form):
...
+    
+    def get_resource_objects(self):
+        self.kwargs[self.user_kw] = self.request.POST[self.user_kw],
+        self.kwargs[self.user_kw] = self.kwargs[self.user_kw][0]
+        self.kwargs[self.module_kw] = self.request.POST[self.module_kw]  
+        super().get_resource_objects()
+
+        #getting module and user by id.
+        self.module = get_object_or_404(
+                        CourseModule,
```
Comment (Mankro) [inline exercise/staff_views.py:573]: Bad indentation.

**Review by Mankro — CHANGES_REQUESTED** [review]
Nice work with your first feature!

I will fix the issues in this pull request myself due to time constraints. I commented on some issues below. There are coding style issues, some significant bugs, lacking access control and user interface improvements.

You could have made more unit tests than just a single very simple test.

**Comment by Mankro** [comment]
The merged commit https://github.com/apluslms/a-plus/commit/7c65106714abe63caa1dfe644adf5ab34d3a75c9 implements the basic submission approval in the inspect view: a single submission or all submissions in one assignment or module.

Approving when adding deviations is part of the unfinished draft pull request #1029.

**Pull request closed without merging**